polyfill/lib/calendar.mjs

```diff
@@ -789,10 +789,32 @@ const nonIsoHelperBase = {
   addMonthsCalendar(calendarDate, months, overflow, cache) {
     const { day } = calendarDate;
     for (let i = 0, absMonths = MathAbs(months); i < absMonths; i++) {
-      const days = months < 0 ? -this.daysInPreviousMonth(calendarDate, cache) : this.daysInMonth(calendarDate, cache);
+      const { month } = calendarDate;
+      const oldCalendarDate = calendarDate;
+      const days =
+        months < 0
+          ? -Math.max(day, this.daysInPreviousMonth(calendarDate, cache))
+          : this.daysInMonth(calendarDate, cache);
       const isoDate = this.calendarToIsoDate(calendarDate, 'constrain', cache);
-      const addedIso = this.addDaysIso(isoDate, days, cache);
+      let addedIso = this.addDaysIso(isoDate, days, cache);
       calendarDate = this.isoToCalendarDate(addedIso, cache);
+
+      // Normally, we can advance one month by adding the number of days in the
+      // current month. However, if we're at the end of the current month and
+      // the next month has fewer days, then we rolled over to the after-next
+      // month. Below we detect this condition and back up until we're back in
+      // the desired month.
+      while (
+        months > 0 &&
```

Comment:
I think this could happen if months < 0 as well? (e.g. November 1 − 30 days = October 2)

Reply:
I don't think so, at least not for Gregorian Nov 1 minus P1M. Following the code above:
* days = -30 (`Math.max(day, this.daysInPreviousMonth(calendarDate, cache))`)
* isoDate = Nov 1
* addedIso = Oct 1 (30 days before Nov 1)
* calendarDate = Oct 1

The logic is a little convoluted here because moving backward one month only requires knowing the length of one month (the previous one) while moving forward one month requires knowing the length of two months: the current one and the next one. (The length of the next one is needed to avoid overflowing the end of the next month, e.g. Jan 31 + P1M. That was #1760.)

It'd make the logic simpler to calculate all month-length values up-front, which would remove the need for this loop because `days` could be calculated precisely for forward adds too. But I didn't want to take the time today to understand the perf implications of calculating two month lengths every time a month is added. The current logic in this PR only pays extra cost on forward adds when wrapping around the end of the year, so only 8% of the time. That seems good enough to me. 

There's *lots* of opportunity for optimizing non-ISO calendars, but honestly I'd prefer someone with better skills than me to tackle that work, and to tackle it in another polyfill. I consider the current non-ISO logic here to be a placeholder that should be produce correct results (hence this PR) but making it really fast is IMHO out of scope here.

BTW, there *was* a bug before this PR with moving backwards, e.g. Mar 31 minus P1M. Current code sets `days` to the length of the previous month (28 days) which is then constrained back to Mar 31. The `Math.max` fixes that bug.

Reply:
_Oops. Edited the comment above with the actual number of days in October. 😄_

Reply:
OK. I do find this logic a bit confusing but I agree with it being a placeholder. That said, would this work?
```js
if (months > 0) {
  const monthsInOldYear = this.monthsInYear(oldCalendarDate, cache);
  while (calendarDate.month - 1 !== month % monthsInOldYear) {
    addedIso = this.addDaysIso(addedIso, -1, cache);
    calendarDate = this.isoToCalendarDate(addedIso, cache);
  }
}
```
(I found it extra confusing because the condition of the `while` loop included some stuff that didn't change from one iteration to the next.)

Reply:
Done!  Nice trick with `%`, although I guess not surprising given that you're a charter member of The Mod Squad. 😄